### DISCUSSION TITLE
Do not use WidgetRef across async gaps?

### DISCUSSION THREAD

**Discussion by stephane-archer** [rrousselGit/riverpod#3859] [Q&A]
We all know about `Do not use BuildContexts across async gaps` but Riverpod uses `WidgetRef` instead of `BuildContexts` like Provider. 

Do we need to care about async gaps with Riverpod too?

``` Dart
class MyWidget extends ConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return ElevatedButton(
      onPressed: () async {
        final result = await fetchData();
        ref.read(myProvider.notifier).update(result); // is it save?
      },
      child: Text('Fetch Data'),
    );
  }
}
```
what about:
``` Dart
onPressed: () async {
        var notifier = ref.read(myProvider.notifier);
        final result = await fetchData();
        notifier.update(result); // is it save?
      },
```

**Comment by rrousselGit** [comment]
WidgetRef == BuildContext.
Yes  you need to be careful about using WidgetRef after async gaps

**Reply by stephane-archer** [reply]
So is it good practice to do ref.read as soon as possible then pass the notifiers as function parameters? What's the best practices ?
I didn't saw any linter complaints, is my linter not correctly configure or is it inexistant ?
If WidgetRef == BuildContext then what's the point of WidgetRef? Is it only to support command line arguments ?

**Reply by rrousselGit** [reply]
WidgetRef is a way for Riverpod to add extra methods to BuildContext, and to keep the API between consumers and providers similar.

> I didn't saw any linter complaints, is my linter not correctly configure or is it inexistant ?

The Flutter tooling has no clue what a WidgetRef is, so there's no warning. Maybe we'll have one in the future.


> So is it good practice to do ref.read as soon as possible then pass the notifiers as function parameters? What's the best practices ?

There's no final answer here. It depends on what your button can or can't do. But if you can start your operation before `await`, then that's probably better yes.

**Reply by stephane-archer** [reply]
It sounds like the provider package is safer to use because it integrates with Flutter tooling.

It's not realistic to be able to do all the operations necessary on providers first, usually, there is an async function to get some data to be able to do the update. In that case, should I access the notifier at the last minute or as soon as the function starts?
